### PR TITLE
feat(pipelines): allow disabling use of change sets

### LINKED ISSUE #20827: (pipelines): Allow to make deployments directly without first creating changeset.
### Describe the feature

Currently Pipelines creates first Changeset, and makes the actual deployment by executing it. If everything goes fine, this is OK and increases the execution time just a bit.

If there's some failure, that is random (e.g. some service used produces timeout, or custom resources in general fail), whole pipeline needs to be rerun as Changeset can't be executed twice.

This would help with e.g. #11995

### Use Case

Currently we run database migrations and trigger test runs (that need to run on target account) as custom resource, but in both of those failure (especially in latter environments) is really time consuming due the need to rerun whole pipeline again.

### Proposed Solution

As Graph implementation offers way to do deployment without prepare -step (https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/pipelines/lib/helpers-internal/pipeline-graph.ts#L28-L34), it would be nice that pipelines would support that (https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/pipelines/lib/codepipeline/codepipeline.ts#L378-L381) same way as Github (https://github.com/cdklabs/cdk-pipelines-github/blob/902e38503e2a5809154cfb574b2a5c7e1798d4af/src/pipeline.ts#L224-L228) does. Preferably with some props, and best would be possibility to define that way per stage or stack.

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.28.1

### Environment details (OS name and version, etc.)

Debian GNU/Linux 11 (bullseye)

### PR DESCRIPTION
**Pull request by Hi-Fi**
Possibility to change deployment way for whole pipeline, for single stage or for single stack in stage.

Fixes #20827

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by Hi-Fi** [comment]
@rix0rrr Does this looks something you thought at the comment at the issue?

File: packages/@aws-cdk/pipelines/README.md [hunk 1]
```diff
@@ -144,6 +144,21 @@ new MyPipelineStack(this, 'PipelineStack', {
 });
 ```
 
+Deployment is done by default with `CodePipeline` engine by first generating
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/README.md:147]: Is this the most logical place in this README to add this content? I feel a separate section might be better.

File: packages/@aws-cdk/pipelines/README.md [hunk 2]
```diff
@@ -144,6 +144,21 @@ new MyPipelineStack(this, 'PipelineStack', {
...
+change set, and then executing it. This allow to add e.g. approval before
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/README.md:148]: Please proof-read your README section once more before submitting.

```suggestion
a change set, and then executing it. This allows you to add e.g. approval before 
```

----------

The reason we have changeset + execute is not (just) for approval steps. Like I said in a comment to the original issue, certain CloudFormation features require a ChangeSet. Hence, it's only fair to list here what features you will be unable to use if you disable the prepare step.

File: packages/@aws-cdk/pipelines/README.md [hunk 3]
```diff
@@ -144,6 +144,21 @@ new MyPipelineStack(this, 'PipelineStack', {
...
+deployment, as change set can't be re-generated if execution of it fails for
+any reason. Deployment can be done as single step by setting 
+`CodePipelineProps.prepareStep` to false (for whole pipeline). It's also 
+possible to change default functionality per stage (with 
+`StageOptions.prepareStep`) or per stack in stage (with
+`StageOptions.prepareStepForStacks`). See examples from tests:
+
+- [whole pipeline](./test/integ.pipeline-without-prepare.ts)
+- [single stage](./test/integ.pipeline-stage-without-prepare.ts)
+- [single stack in stage](./test/integ.pipeline-stage-stack-without-prepare.ts)
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/README.md:159]: These relative hyperlinks will only work on GitHub. 

* They do not work here: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.pipelines-readme.html
* They do not work here: https://www.npmjs.com/package/@aws-cdk/pipelines

Instead, put in a code example directly that showcases how to use this feature in a simple way. It's okay to not cover every permutation.

File: packages/@aws-cdk/pipelines/README.md [hunk 4]
```diff
@@ -867,11 +882,11 @@ because those values are passed in directly to the underlying `codepipeline.Pipe
 
 Docker can be used in 3 different places in the pipeline:
 
-* If you are using Docker image assets in your application stages: Docker will
+- If you are using Docker image assets in your application stages: Docker will
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/README.md:885]: Why this change?

File: packages/@aws-cdk/pipelines/lib/helpers-internal/pipeline-graph.ts [hunk 5]
```diff
@@ -333,7 +346,7 @@ type GraphAnnotation =
   | { readonly type: 'step'; readonly step: Step; isBuildStep?: boolean }
   | { readonly type: 'self-update' }
   | { readonly type: 'prepare'; readonly stack: StackDeployment }
-  | { readonly type: 'execute'; readonly stack: StackDeployment; readonly captureOutputs: boolean }
+  | { readonly type: 'execute'; readonly stack: StackDeployment; readonly captureOutputs: boolean; readonly withoutChangeset?: boolean }
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/lib/helpers-internal/pipeline-graph.ts:349]: These types constitute a contract, that is consumed for example by https://github.com/cdklabs/cdk-pipelines-github

Changing what the `execute` node means will break other implementors. Instead, add a new node type (for example, `deploy`) that has a new meaning.

**Review by rix0rrr — DISMISSED** [review]
Thanks! I have a request for a couple of changes, see below.

File: packages/@aws-cdk/pipelines/README.md (hunk 4 above)
Comment (Hi-Fi) [reply packages/@aws-cdk/pipelines/README.md:885]: This came automatically with markdownlint.

File: packages/@aws-cdk/pipelines/README.md (hunk 2 above)
Comment (Hi-Fi) [reply packages/@aws-cdk/pipelines/README.md:148]: Is there somewhere list of those or documentation about features that require change set? Dynamic referring is only one mentioned in issue.

File: packages/@aws-cdk/pipelines/lib/helpers-internal/pipeline-graph.ts (hunk 5 above)
Comment (Hi-Fi) [reply packages/@aws-cdk/pipelines/lib/helpers-internal/pipeline-graph.ts:349]: Would it break, of there is new optional field? I first checked separate one, but at least thst point it seemed to generate more changes.

Of course now with working tests it would be easier to refactor.

Comment (Hi-Fi) [reply packages/@aws-cdk/pipelines/lib/helpers-internal/pipeline-graph.ts:349]: I think that adding new node here is easier breaking change, as node is created at https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/pipelines/lib/helpers-internal/pipeline-graph.ts#L138-L142.

Here if we determine the new `deploy` node if no prepare step is needed, existing functionality would fail as e.g. GitHub engine would get different node what it can't handle at https://github.com/cdklabs/cdk-pipelines-github/blob/3fa9bb2df734c4eb6b5ce05946e728af65881037/src/pipeline.ts#L350-L351.

Or did I misunderstood? Optional prop should be invisible to all existing implementors, and allows those to continue as is.

Of course that execute node in general is a bit confusing, as with that existing logic to leave out prepare step single step is not really execute but deploy https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/pipelines/lib/helpers-internal/pipeline-graph.ts#L137.

File: packages/@aws-cdk/pipelines/README.md [hunk 6]
```diff
@@ -544,6 +544,79 @@ const pipeline = new pipelines.CodePipeline(this, 'Pipeline', {
 });
 ```
 
+#### Deploying without change sets
+
+Deployment is done by default with `CodePipeline` engine using change sets, 
+i.e. to first create a change set and then execute it. This allows you to:
+
+- have approval before actual deployment happens
+- include [dynamic references](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references.html)
```
Comment (polothy) [inline packages/@aws-cdk/pipelines/README.md:553]: From the linked doc:

> When you use a dynamic reference, CloudFormation retrieves the value of the specified reference when necessary during _stack_ and change set operations.

Sort of sounds like normal stack operations also support them.  Was anyone able to try to see if it's a problem (or past experience)? I fully realize that the docs just might not reflect the reality of things.

Comment (polothy) [reply packages/@aws-cdk/pipelines/README.md:553]: Oh gees, this added me as a reviewer, sorry. Please feel free to ignore.

Comment (Hi-Fi) [reply packages/@aws-cdk/pipelines/README.md:553]: I think one example that can't be done with normal stack is that you have some value in secrets manager that you want to use at created resource, but resource itself doesn't support references. So e.g. with CodeBuild and ECS you can refer to secrets manager value with ARN, but if you make e.g. OIDC authentication at ALB level, you can only provide strings. With changeset it's possible to get the secret to ALB without having it in the template

**Review by rix0rrr — APPROVED** [review]
I see what you're saying on the 'execute' step. It's a bit weird, but apparently we're already doing this for the GitHub provider.

I've simplified the PR a little: allow disabling use of changesets altogether for the `CodePipeline` provider (and not customize on stage or stack level). I'm told I'm wrong and direct deployments should just work, so the only reason to have changesets might just be to add an approval step.

If we want to go the "customize per stack" route, we'll have to do it slightly differently anyway. Currently `prepareStep: false` as passed by the GitHub engine doesn't mean "I don't **want** change sets", it means "I don't **support** change sets".

Since the per-stage and per-stack API is shared between all Pipelines implementations, it would conceivably be possible to turn it **on** on a per-stage basis, which should lead to an error for the GitHub pipeline for which that is not possible.

File: packages/@aws-cdk/pipelines/README.md (hunk 2 above)
Comment (fab-mindflow) [reply packages/@aws-cdk/pipelines/README.md]: The list has not been added to documentation.
Below in this review, @rix0rrr, you say:

>  I'm told I'm wrong and direct deployments should just work, so the only reason to have changesets might just be to add an approval step.

Is this confirmed that this is the only difference?
(we are using a 30 stacks pipeline, without need for approval step, and saving 1 minute per stack when they are not updated is enormous - it seems to work without any side-effect for now)
